sqlmesh/core/plan/builder.py

```diff
@@ -650,6 +709,85 @@ def _ensure_new_env_with_changes(self) -> None:
                 "No changes were detected. Make a change or run with --include-unmodified to create a new environment without changes."
             )
 
+    def _upstream_directly_modified(
+        self, s_id: SnapshotId, directly_modified: t.Set[SnapshotId], dag: DAG[SnapshotId]
+    ) -> t.List[SnapshotId]:
+        return [
+            upstream_id for upstream_id in dag.upstream(s_id) if upstream_id in directly_modified
+        ]
+
+    def _validate_schema_change(
+        self,
+        s_id: SnapshotId,
+        upstream_snapshot_ids: t.List[SnapshotId] = [],
+        force_warn: bool = False,
+    ) -> None:
+        """Determine if a schema change requires a DROP operation"""
+        s_id_snapshot = t.cast(Snapshot, self._context_diff.snapshots.get(s_id))
+
+        if not s_id_snapshot.model.on_schema_change.is_ignore or force_warn:
+            schema_differ_args = getattr(
+                DIALECT_TO_ENGINE_ADAPTER[s_id_snapshot.model.dialect], "schema_differ_args"
+            )
+            schema_differ = SchemaDiffer(**schema_differ_args)
+
+            subdag = [s_id, *upstream_snapshot_ids]
+
+            # if we already know that the change is destructive, we don't need to evaluate any more snapshots
+            warning_msg = f"PLAN TIME CHECK: Plan results in a destructive change to forward-only model '{s_id_snapshot.name}'s schema."
+            error_msg = f"{warning_msg} To allow this, change the model's `on_schema_change` setting to `warn` or `ignore` or include it in the plan's `--allow-destructive-model` option."
+
+            if any(self._snapshot_change_is_destructive.get(id.name, None) for id in subdag):
```

Comment:
you're using any a lot, is there another way we can do this or another data structure?